### ISSUE TITLE
Nested regexes cause issues in Lint/OutOfRangeRegexRef

### ISSUE THREAD

**Issue by tas50** [rubocop/rubocop#9269]
## Expected behavior

Lint/OutOfRangeRegexRef would not alert when using a ref that is actually valid/present.

## Actual behavior

RuboCop incorrectly identifies the outer regex as being an issue due to the lack of that ref in the inner. It's a bit funky so check the example to see what I meant here.

## Steps to reproduce the problem

Given this code:

```ruby
            case line
              when /^#{Regexp.escape(var_name)}="(\w+)"/
                enabled_state_found!
                if $1 =~ /^yes$/i
                  current_resource.enabled true
                elsif $1 =~ /^(no|none)$/i
                  current_resource.enabled false
                end
              end
```

RuboCop is matching on this part:

```
Lint/OutOfRangeRegexpRef: $1 is out of range (no regexp capture groups detected).
                elsif $1 =~ /^(no|none)$/i
                      ^^
```

The $1 does exist and is being compared with =~. In this case RuboCop should not alert.

## RuboCop version

Current master

```
1.6.1 (using Parser 2.7.2.0, rubocop-ast 1.3.0, running on ruby 2.7.2 x86_64-darwin19)
  - rubocop-performance 1.9.1
  - rubocop-rspec 2.0.1
```

**Comment by jonas054** [comment]
The offense report is correct. Consider this slightly modified code:
```ruby
var_name = 'foo'
line = 'foo="no"'

case line
when /^#{Regexp.escape(var_name)}="(\w+)"/
  p before: $1
  if $1 =~ /^yes$/i
    p 'yes'
  else
    p after: $1
    if $1 =~ /^(no|none)$/i
      p 'no'
    end
  end
end
```
It outputs:
```
{:before=>"no"}
{:after=>nil}
```
That's because `$1` is reset when `if $1 =~ /^yes$/i` is executed.

**Comment by bbatsov** [comment]
Closing — a collaborator explained the offense report is actually correct for the nested regex case. No further discussion in 5+ years.
